x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/RetryingHttpSender.java

```diff
@@ -108,23 +108,29 @@ public void tryAction(ActionListener<InferenceServiceResults> listener) {
                 return;
             }
 
-            ActionListener<HttpResult> responseListener = ActionListener.wrap(result -> {
-                try {
-                    responseHandler.validateResponse(throttlerManager, logger, request, result);
-                    InferenceServiceResults inferenceResults = responseHandler.parseResult(request, result);
-
-                    listener.onResponse(inferenceResults);
-                } catch (Exception e) {
-                    logException(logger, request, result, responseHandler.getRequestType(), e);
-                    listener.onFailure(e);
-                }
-            }, e -> {
+            var retryableListener = listener.delegateResponse((l, e) -> {
                 logException(logger, request, responseHandler.getRequestType(), e);
-                listener.onFailure(transformIfRetryable(e));
+                l.onFailure(transformIfRetryable(e));
             });
 
             try {
-                httpClient.send(request.createHttpRequest(), context, responseListener);
+                if (request.isStreaming() && responseHandler.canHandleStreamingResponses()) {
+                    httpClient.stream(request.createHttpRequest(), context, retryableListener.delegateFailure((l, r) -> {
+                        r.subscribe(new StreamingResponseHandler(throttlerManager, logger, request, responseHandler, l));
+                    }));
+                } else {
+                    httpClient.send(request.createHttpRequest(), context, retryableListener.delegateFailure((l, r) -> {
```

Comment:
If is adds complications then I can make a case for not supporting retry with streaming completions.

Retry is useful for an unattended bulk operation such as embeddings at ingest, it's better to retry and overcome transient issues. Streaming completions is attended, someone is likely reading the response if they want to retry they can press the button again. Same with a client.

Reply:
We would really only process the retry before the connection and stream are established - if the connection is interrupted mid-stream or something, we currently do not retry, we'll just forward the error to the user